pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package kademlia
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/addressbook"
+	"github.com/ethersphere/bee/pkg/discovery"
+	"github.com/ethersphere/bee/pkg/kademlia/pslice"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	maxBins        = 16
```

Comment:
there's no max for now, this is a complicated issue, i explained this thoroughly in the `doc.go` file

Reply:
In the doc your primary argument is that it is cheap to maintain connections, which is not true.
These are expected to be long standing connections and because of the non-indexed PSLice implementation, can become a performance bottleneck in future. Also the more the peers in the closest bins, the more bandwidth you spend to share chunks using pull sync.

Reply:
> which is not true

i argue that _it should be_.
you are welcome to explain also why at this point it is not cheap to maintain open connections to peers which are normally not used. we will have to design pull sync so that these bottlenecks do not occur, and i have this very high up on my priority list. 

maintaining connections _should be_ cheap. we should design around the fact that we could have a better connectivity, not that we need to limit our connectivity because our software is designed in the wrong way that cannot maintain connectivity because it overloads itself.

Reply:
> you are welcome to explain also why at this point it is not cheap to maintain open connections to peers

Well i would not take this comment  further if you think so. If maintaining TCP connections are cheap people would have built system which would connect to thousands of node and p2p would not have born. 

Also at this stage we don't have a max peer limit after which we reject peer connections. So if someone decides to bring a node down, he has to generate lost of addresses which are 1 bit apart and keep sending to you and you will happily accept it until you reach 65535 or 300+ in windows.